code/game/machinery/doors/firedoor.dm

```diff
@@ -44,29 +93,131 @@
-	remove_from_areas()
-	affecting_areas.Cut()
-	return ..()
+	start_activation_process(result)
+
+/**
+ * Begins activation process of us and our neighbors.
+ *
+ * This proc will call activate() on every fire lock (including us) listed
+ * in the merge group datum. Returns without doing anything if our alarm_type
+ * was already set, as that means that we're already active.
+ *
+ * Arguments:
+ * code should be one of three defined alarm types, or can be not supplied. Will dictate the color of the fire alarm lights, and defults to "firelock_alarm_type_generic"
+ */
+/obj/machinery/door/firedoor/proc/start_activation_process(code = FIRELOCK_ALARM_TYPE_GENERIC)
+	if(alarm_type)
+		return //We're already active
+	var/datum/merger/merge_group = GetMergeGroup(merger_id, merger_typecache)
+	for(var/obj/machinery/door/firedoor/buddylock as anything in merge_group.members)
+		buddylock.activate(code)
+
+
+/**
+ * Proc that handles activation of the firelock and all this details
+ *
+ * Sets the alarm_type variable based on the single arg, which is in turn
+ * used by several procs to understand the intended state of the fire lock.
+ * Also calls set_status() on all fire alarms in all affected areas, tells
+ * the area the firelock sits in to report the event (AI, alarm consoles, etc)
+ * and finally calls correct_state(), which will handle opening or closing
+ * this fire lock.
+ */
+/obj/machinery/door/firedoor/proc/activate(code = FIRELOCK_ALARM_TYPE_GENERIC)
+	SIGNAL_HANDLER
+	if(alarm_type)
+		return //Already active
+	alarm_type = code
+	for(var/area/place in affecting_areas)
+		LAZYADD(place.active_firelocks, src)
+		if(LAZYLEN(place.active_firelocks) == 1) //if we're the first to activate in this particular area
+			for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+				fire_panel.set_status()
+			if(place == my_area)
+				place.alarm_manager.send_alarm(ALARM_FIRE, place) //We'll limit our reporting to just the area we're on. If the issue affects bordering areas, they can report it themselves
+	update_icon() //Sets the door lights even if the door doesn't move.
+	correct_state()
+
+/**
+ * Proc that handles reset steps
+ *
+ * Clears the alarm state and attempts to open the firelock.
+ */
+/obj/machinery/door/firedoor/proc/reset()
+	SIGNAL_HANDLER
+	alarm_type = null
+	for(var/area/place in affecting_areas)
+		LAZYREMOVE(place.active_firelocks, src)
+		if(!LAZYLEN(place.active_firelocks)) //if we were the last firelock still active in this particular area
+			for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+				fire_panel.set_status()
+	COOLDOWN_START(src, detect_cooldown, DETECT_COOLDOWN_STEP_TIME)
+	update_icon() //Sets the door lights even if the door doesn't move.
+	correct_state()
+
+/obj/machinery/door/firedoor/emag_act(mob/user, obj/item/card/emag/doorjack/digital_crowbar)
+	if(!istype(digital_crowbar))
+		return //emags get out
+	if(obj_flags &= EMAGGED || !user || digital_crowbar.charges < 1)
```

Comment:
why are you doing &=?

Reply:
Typo. Or maybe was dumb. I don't remember. Has been fixed.